plugins/out_cloudwatch_logs/cloudwatch_api.c

```diff
@@ -0,0 +1,1072 @@
+}
+
+/* sets the oldest_event and newest_event fields */
+static void set_stream_time_span(struct log_stream *stream, struct event *event)
+{
+    if (stream->oldest_event == 0 || stream->oldest_event > event->timestamp) {
+        stream->oldest_event = event->timestamp;
+    }
+
+    if (stream->newest_event == 0 || stream->newest_event < event->timestamp) {
+        stream->newest_event = event->timestamp;
+    }
+}
+
+
+/*
+ * Processes the msgpack object
+ * Returns 0 on success, -1 on general errors,
+ * and 1 if we ran out of space to write the event
+ * which means a send must occur
+ */
+int process_event(struct flb_cloudwatch *ctx, struct cw_flush *buf,
+                  const msgpack_object *obj, struct flb_time *tms)
+{
+    size_t written;
+    size_t size;
+    int offset = 0;
+    struct event *event;
+    char *tmp_buf_ptr;
+
+    tmp_buf_ptr = buf->tmp_buf + buf->tmp_buf_offset;
+    written = flb_msgpack_to_json(tmp_buf_ptr,
+                                  buf->tmp_buf_size - buf->tmp_buf_offset,
+                                  obj);
+    if (written < 0) {
+        /*
+         * negative value means failure to write to buffer,
+         * which means we ran out of space, and must send the logs
+         */
+        return 1;
+    }
+    /* Discard empty messages (written == 2 means '""') */
+    if (written <= 2) {
+        flb_plg_debug(ctx->ins, "Found empty log message");
+        return 0;
+    }
+
+    /* the json string must be escaped, unless the log_key option is used */
+    if (ctx->log_key == NULL) {
+        /*
+         * check if event_buf is initialized and big enough
+         * If all chars need to be hex encoded (impossible), 6x space would be
+         * needed
+         */
+        size = written * 6;
+        if (buf->event_buf == NULL || buf->event_buf_size < size) {
+            flb_free(buf->event_buf);
+            buf->event_buf = flb_malloc(size);
+            buf->event_buf_size = size;
+            if (buf->event_buf == NULL) {
+                flb_errno();
+                return -1;
+            }
+        }
+        offset = 0;
+        if (!flb_utils_write_str(buf->event_buf, &offset, size,
+                                 tmp_buf_ptr, written)) {
+            return -1;
+        }
+        written = offset;
+
+        tmp_buf_ptr = buf->tmp_buf + buf->tmp_buf_offset;
+        if ((buf->tmp_buf_size - buf->tmp_buf_offset) < written) {
+            /* not enough space, send logs */
+            return 1;
+        }
+
+        /* copy serialized json to tmp_buf */
+        if (!strncpy(tmp_buf_ptr, buf->event_buf, written)) {
+            return -1;
+        }
+
+        buf->tmp_buf_offset += written;
+        event = &buf->events[buf->event_index];
+        event->json = tmp_buf_ptr;
+        event->len = written;
+        event->timestamp = (unsigned long long) (tms->tm.tv_sec * 1000 +
+                                                 tms->tm.tv_nsec/1000000);
+
+    } else {
```

Comment:
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions